{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "4fbb9533",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/data/yangz/bioinfo/packages/exprmat/exprmat/data/mmu/assemblies/grcm39'"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "os.chdir('/home/data/yangz/bioinfo/packages/exprmat/exprmat/data/mmu/assemblies/grcm39')\n",
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "c8f56fc0",
   "metadata": {},
   "outputs": [],
   "source": [
    "ann = pd.read_table('annotations.gff.gz', comment = '#', sep = '\\t', header = None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "e2e8e130",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>3143476</td>\n",
       "      <td>3144545</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSMUSG00000102693.2;gene_id=ENSMUSG0000010...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3143476</td>\n",
       "      <td>3144545</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSMUST00000193812.2;Parent=ENSMUSG00000102...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>3143476</td>\n",
       "      <td>3144545</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENSMUST00000193812.2:1;Parent=ENSMUST0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>3172239</td>\n",
       "      <td>3172348</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSMUSG00000064842.3;gene_id=ENSMUSG0000006...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3172239</td>\n",
       "      <td>3172348</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSMUST00000082908.3;Parent=ENSMUSG00000064...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1312349</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15289</td>\n",
       "      <td>15355</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSMUST00000082422.1;Parent=ENSMUSG00000064...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1312350</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>15289</td>\n",
       "      <td>15355</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENSMUST00000082422.1:1;Parent=ENSMUST0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1312351</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>15356</td>\n",
       "      <td>15422</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSMUSG00000064372.1;gene_id=ENSMUSG0000006...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1312352</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15356</td>\n",
       "      <td>15422</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSMUST00000082423.1;Parent=ENSMUSG00000064...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1312353</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>15356</td>\n",
       "      <td>15422</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENSMUST00000082423.1:1;Parent=ENSMUST0...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1312354 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            0        1           2        3        4  5  6  7  \\\n",
       "0        chr1   HAVANA        gene  3143476  3144545  .  +  .   \n",
       "1        chr1   HAVANA  transcript  3143476  3144545  .  +  .   \n",
       "2        chr1   HAVANA        exon  3143476  3144545  .  +  .   \n",
       "3        chr1  ENSEMBL        gene  3172239  3172348  .  +  .   \n",
       "4        chr1  ENSEMBL  transcript  3172239  3172348  .  +  .   \n",
       "...       ...      ...         ...      ...      ... .. .. ..   \n",
       "1312349  chrM  ENSEMBL  transcript    15289    15355  .  +  .   \n",
       "1312350  chrM  ENSEMBL        exon    15289    15355  .  +  .   \n",
       "1312351  chrM  ENSEMBL        gene    15356    15422  .  -  .   \n",
       "1312352  chrM  ENSEMBL  transcript    15356    15422  .  -  .   \n",
       "1312353  chrM  ENSEMBL        exon    15356    15422  .  -  .   \n",
       "\n",
       "                                                         8  \n",
       "0        ID=ENSMUSG00000102693.2;gene_id=ENSMUSG0000010...  \n",
       "1        ID=ENSMUST00000193812.2;Parent=ENSMUSG00000102...  \n",
       "2        ID=exon:ENSMUST00000193812.2:1;Parent=ENSMUST0...  \n",
       "3        ID=ENSMUSG00000064842.3;gene_id=ENSMUSG0000006...  \n",
       "4        ID=ENSMUST00000082908.3;Parent=ENSMUSG00000064...  \n",
       "...                                                    ...  \n",
       "1312349  ID=ENSMUST00000082422.1;Parent=ENSMUSG00000064...  \n",
       "1312350  ID=exon:ENSMUST00000082422.1:1;Parent=ENSMUST0...  \n",
       "1312351  ID=ENSMUSG00000064372.1;gene_id=ENSMUSG0000006...  \n",
       "1312352  ID=ENSMUST00000082423.1;Parent=ENSMUSG00000064...  \n",
       "1312353  ID=exon:ENSMUST00000082423.1:1;Parent=ENSMUST0...  \n",
       "\n",
       "[1312354 rows x 9 columns]"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ann"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "80649af8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2\n",
       "exon                                      536952\n",
       "CDS                                       428695\n",
       "transcript                                 84484\n",
       "five_prime_UTR                             69039\n",
       "gene                                       56691\n",
       "three_prime_UTR                            45584\n",
       "start_codon                                45536\n",
       "stop_codon                                 45315\n",
       "stop_codon_redefined_as_selenocysteine        58\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ann[2].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "c9214783",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "1312354it [00:08, 152979.48it/s]\n"
     ]
    }
   ],
   "source": [
    "id = []\n",
    "gene_id = []\n",
    "subtype = []\n",
    "parent = []\n",
    "mapper = {\n",
    "    'exon': 'exon',\n",
    "    'UTR5': 'utr5',\n",
    "    'UTR3': 'utr3',\n",
    "    'CDS': 'cds',\n",
    "    'UTR': 'utr',\n",
    "    'stop_codon': 'stop',\n",
    "    'start_codon': 'start',\n",
    "    'selenocysteine': 'stop.sel'\n",
    "}\n",
    "\n",
    "from tqdm import tqdm \n",
    "for line, recordtype in tqdm(zip(ann[8].tolist(), ann[2].tolist())):\n",
    "    terms = line.split(';')\n",
    "    _id = ''\n",
    "    _gene_id = ''\n",
    "    _subtype = ''\n",
    "    _parent = ''\n",
    "    for term in terms:\n",
    "        pair = term.split('=')\n",
    "        \n",
    "        if pair[0] == 'ID':\n",
    "            if (\n",
    "                pair[1].startswith('exon:') or\n",
    "                pair[1].startswith('UTR5:') or\n",
    "                pair[1].startswith('UTR3:') or\n",
    "                pair[1].startswith('CDS:') or\n",
    "                pair[1].startswith('UTR:') or\n",
    "                pair[1].startswith('stop_codon:') or\n",
    "                pair[1].startswith('start_codon:') or\n",
    "                pair[1].startswith('selenocysteine:')\n",
    "            ): \n",
    "                stems = pair[1].split(':')\n",
    "                if len(stems) == 2:\n",
    "                    _id = mapper[stems[0]] + ':1'\n",
    "                else: _id = mapper[stems[0]] + ':' + stems[2]\n",
    "            \n",
    "            else: _id = pair[1]\n",
    "        \n",
    "        elif pair[0] == 'gene_id':\n",
    "            _gene_id = pair[1]\n",
    "\n",
    "        elif pair[0] == 'gene_type' and recordtype == 'gene':    \n",
    "            _subtype = pair[1]\n",
    "        \n",
    "        elif pair[0] == 'transcript_type' and recordtype == 'transcript':\n",
    "            _subtype = pair[1]\n",
    "        \n",
    "        elif pair[0] == 'Parent':\n",
    "            _parent = pair[1]\n",
    "    \n",
    "    id.append(_id)\n",
    "    gene_id.append(_gene_id)\n",
    "    subtype.append(_subtype)\n",
    "    parent.append(_parent)\n",
    "\n",
    "hier = pd.DataFrame({\n",
    "    'id': id,\n",
    "    'gid': gene_id,\n",
    "    'subtype': subtype,\n",
    "    'parent': parent\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "e862313c",
   "metadata": {},
   "outputs": [],
   "source": [
    "hier['id'] = [x.split('.')[0] for x in hier['id']]\n",
    "hier['gid'] = [x.split('.')[0] for x in hier['gid']]\n",
    "hier['parent'] = [x.split('.')[0] for x in hier['parent']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "69b21c6b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hier['id'].is_unique"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "c1d36320",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged = pd.concat([\n",
    "    ann[[0, 2, 3, 4, 6]],\n",
    "    hier\n",
    "], axis = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "948bc5b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged['chr'] = ['MT' if x == 'chrM' else x.replace('chr', '') for x in merged[0]]\n",
    "merged.columns = [\n",
    "    'ucsc', 'type', 'start', 'end', 'strand',\n",
    "    'id', 'gid', 'subtype', 'parent', 'chr'\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "f1711d0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged.to_feather('gene-models.feather')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "4eeb1c47",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ucsc</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>id</th>\n",
       "      <th>gid</th>\n",
       "      <th>subtype</th>\n",
       "      <th>parent</th>\n",
       "      <th>chr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>gene</td>\n",
       "      <td>3143476</td>\n",
       "      <td>3144545</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUSG00000102693</td>\n",
       "      <td>ENSMUSG00000102693</td>\n",
       "      <td>TEC</td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3143476</td>\n",
       "      <td>3144545</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000193812</td>\n",
       "      <td>ENSMUSG00000102693</td>\n",
       "      <td>TEC</td>\n",
       "      <td>ENSMUSG00000102693</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>3143476</td>\n",
       "      <td>3144545</td>\n",
       "      <td>+</td>\n",
       "      <td>exon:1</td>\n",
       "      <td>ENSMUSG00000102693</td>\n",
       "      <td></td>\n",
       "      <td>ENSMUST00000193812</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>gene</td>\n",
       "      <td>3172239</td>\n",
       "      <td>3172348</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUSG00000064842</td>\n",
       "      <td>ENSMUSG00000064842</td>\n",
       "      <td>snRNA</td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3172239</td>\n",
       "      <td>3172348</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000082908</td>\n",
       "      <td>ENSMUSG00000064842</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>ENSMUSG00000064842</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1312349</th>\n",
       "      <td>chrM</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15289</td>\n",
       "      <td>15355</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000082422</td>\n",
       "      <td>ENSMUSG00000064371</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>ENSMUSG00000064371</td>\n",
       "      <td>MT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1312350</th>\n",
       "      <td>chrM</td>\n",
       "      <td>exon</td>\n",
       "      <td>15289</td>\n",
       "      <td>15355</td>\n",
       "      <td>+</td>\n",
       "      <td>exon:1</td>\n",
       "      <td>ENSMUSG00000064371</td>\n",
       "      <td></td>\n",
       "      <td>ENSMUST00000082422</td>\n",
       "      <td>MT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1312351</th>\n",
       "      <td>chrM</td>\n",
       "      <td>gene</td>\n",
       "      <td>15356</td>\n",
       "      <td>15422</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000064372</td>\n",
       "      <td>ENSMUSG00000064372</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td></td>\n",
       "      <td>MT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1312352</th>\n",
       "      <td>chrM</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15356</td>\n",
       "      <td>15422</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000082423</td>\n",
       "      <td>ENSMUSG00000064372</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>ENSMUSG00000064372</td>\n",
       "      <td>MT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1312353</th>\n",
       "      <td>chrM</td>\n",
       "      <td>exon</td>\n",
       "      <td>15356</td>\n",
       "      <td>15422</td>\n",
       "      <td>-</td>\n",
       "      <td>exon:1</td>\n",
       "      <td>ENSMUSG00000064372</td>\n",
       "      <td></td>\n",
       "      <td>ENSMUST00000082423</td>\n",
       "      <td>MT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1312354 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         ucsc        type    start      end strand                  id  \\\n",
       "0        chr1        gene  3143476  3144545      +  ENSMUSG00000102693   \n",
       "1        chr1  transcript  3143476  3144545      +  ENSMUST00000193812   \n",
       "2        chr1        exon  3143476  3144545      +              exon:1   \n",
       "3        chr1        gene  3172239  3172348      +  ENSMUSG00000064842   \n",
       "4        chr1  transcript  3172239  3172348      +  ENSMUST00000082908   \n",
       "...       ...         ...      ...      ...    ...                 ...   \n",
       "1312349  chrM  transcript    15289    15355      +  ENSMUST00000082422   \n",
       "1312350  chrM        exon    15289    15355      +              exon:1   \n",
       "1312351  chrM        gene    15356    15422      -  ENSMUSG00000064372   \n",
       "1312352  chrM  transcript    15356    15422      -  ENSMUST00000082423   \n",
       "1312353  chrM        exon    15356    15422      -              exon:1   \n",
       "\n",
       "                        gid  subtype              parent chr  \n",
       "0        ENSMUSG00000102693      TEC                       1  \n",
       "1        ENSMUSG00000102693      TEC  ENSMUSG00000102693   1  \n",
       "2        ENSMUSG00000102693           ENSMUST00000193812   1  \n",
       "3        ENSMUSG00000064842    snRNA                       1  \n",
       "4        ENSMUSG00000064842    snRNA  ENSMUSG00000064842   1  \n",
       "...                     ...      ...                 ...  ..  \n",
       "1312349  ENSMUSG00000064371  Mt_tRNA  ENSMUSG00000064371  MT  \n",
       "1312350  ENSMUSG00000064371           ENSMUST00000082422  MT  \n",
       "1312351  ENSMUSG00000064372  Mt_tRNA                      MT  \n",
       "1312352  ENSMUSG00000064372  Mt_tRNA  ENSMUSG00000064372  MT  \n",
       "1312353  ENSMUSG00000064372           ENSMUST00000082423  MT  \n",
       "\n",
       "[1312354 rows x 10 columns]"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "exprmat.py310.r43.torch",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
