{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "e7fd0361",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/data/yangz/bioinfo/packages/exprmat/exprmat/data/hsa/assemblies/grch37'"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "os.chdir('/home/data/yangz/bioinfo/packages/exprmat/exprmat/data/hsa/assemblies/grch37')\n",
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "40530ac6",
   "metadata": {},
   "outputs": [],
   "source": [
    "root = \"https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/405/GCF_000001405.25_GRCh37.p13/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d38ce0eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "!wget -O gaps.tsv.gz {root}GCF_000001405.25_GRCh37.p13_genomic_gaps.txt.gz\n",
    "!gunzip gaps.tsv.gz\n",
    "!wget -O patches.tsv {root}GCF_000001405.25_GRCh37.p13_assembly_regions.txt\n",
    "!wget -O regions.tsv {root}GCF_000001405.25_GRCh37.p13_assembly_report.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "7cdcf23a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "e9204d3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "gaps = pd.read_table('gaps.tsv', comment = '#', sep = '\\t', header = None)\n",
    "patches = pd.read_table('patches.tsv', comment = '#', sep = '\\t', header = None)\n",
    "regions = pd.read_table('regions.tsv', comment = '#', sep = '\\t', header = None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "6c5c79fb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>.seqid</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>length</th>\n",
       "      <th>type</th>\n",
       "      <th>linkage.evidence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NT_113907.1</td>\n",
       "      <td>23054</td>\n",
       "      <td>23153</td>\n",
       "      <td>100</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unspecified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NT_113950.2</td>\n",
       "      <td>9934</td>\n",
       "      <td>10033</td>\n",
       "      <td>100</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unspecified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NT_167244.1</td>\n",
       "      <td>207808</td>\n",
       "      <td>252060</td>\n",
       "      <td>44253</td>\n",
       "      <td>within_scaffold</td>\n",
       "      <td>clone_contig</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NT_167244.1</td>\n",
       "      <td>327075</td>\n",
       "      <td>356458</td>\n",
       "      <td>29384</td>\n",
       "      <td>within_scaffold</td>\n",
       "      <td>clone_contig</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NT_167244.1</td>\n",
       "      <td>448575</td>\n",
       "      <td>536172</td>\n",
       "      <td>87598</td>\n",
       "      <td>within_scaffold</td>\n",
       "      <td>clone_contig</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>508</th>\n",
       "      <td>NC_000024.9</td>\n",
       "      <td>22369680</td>\n",
       "      <td>22419679</td>\n",
       "      <td>50000</td>\n",
       "      <td>within_scaffold</td>\n",
       "      <td>clone_contig</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>509</th>\n",
       "      <td>NC_000024.9</td>\n",
       "      <td>23901429</td>\n",
       "      <td>23951428</td>\n",
       "      <td>50000</td>\n",
       "      <td>between_scaffolds</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>510</th>\n",
       "      <td>NC_000024.9</td>\n",
       "      <td>28819362</td>\n",
       "      <td>58819361</td>\n",
       "      <td>30000000</td>\n",
       "      <td>heterochromatin</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>511</th>\n",
       "      <td>NC_000024.9</td>\n",
       "      <td>58917657</td>\n",
       "      <td>58967656</td>\n",
       "      <td>50000</td>\n",
       "      <td>between_scaffolds</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>512</th>\n",
       "      <td>NC_000024.9</td>\n",
       "      <td>59363567</td>\n",
       "      <td>59373566</td>\n",
       "      <td>10000</td>\n",
       "      <td>telomere</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>513 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          .seqid     start       end    length               type  \\\n",
       "0    NT_113907.1     23054     23153       100            unknown   \n",
       "1    NT_113950.2      9934     10033       100            unknown   \n",
       "2    NT_167244.1    207808    252060     44253    within_scaffold   \n",
       "3    NT_167244.1    327075    356458     29384    within_scaffold   \n",
       "4    NT_167244.1    448575    536172     87598    within_scaffold   \n",
       "..           ...       ...       ...       ...                ...   \n",
       "508  NC_000024.9  22369680  22419679     50000    within_scaffold   \n",
       "509  NC_000024.9  23901429  23951428     50000  between_scaffolds   \n",
       "510  NC_000024.9  28819362  58819361  30000000    heterochromatin   \n",
       "511  NC_000024.9  58917657  58967656     50000  between_scaffolds   \n",
       "512  NC_000024.9  59363567  59373566     10000           telomere   \n",
       "\n",
       "    linkage.evidence  \n",
       "0        unspecified  \n",
       "1        unspecified  \n",
       "2       clone_contig  \n",
       "3       clone_contig  \n",
       "4       clone_contig  \n",
       "..               ...  \n",
       "508     clone_contig  \n",
       "509               na  \n",
       "510               na  \n",
       "511               na  \n",
       "512               na  \n",
       "\n",
       "[513 rows x 6 columns]"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gaps.columns = ['.seqid', 'start', 'end', 'length', 'type', 'linkage.evidence']\n",
    "gaps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "80526697",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>type</th>\n",
       "      <th>genbank</th>\n",
       "      <th>refseq</th>\n",
       "      <th>unit</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>REGION77</td>\n",
       "      <td>1</td>\n",
       "      <td>3823584.0</td>\n",
       "      <td>3845268.0</td>\n",
       "      <td>fix-patch</td>\n",
       "      <td>JH806574.2</td>\n",
       "      <td>NW_004070864.2</td>\n",
       "      <td>PATCHES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>REGION60</td>\n",
       "      <td>1</td>\n",
       "      <td>31872759.0</td>\n",
       "      <td>32017063.0</td>\n",
       "      <td>fix-patch</td>\n",
       "      <td>GL949741.1</td>\n",
       "      <td>NW_003571030.1</td>\n",
       "      <td>PATCHES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1P11</td>\n",
       "      <td>1</td>\n",
       "      <td>120471050.0</td>\n",
       "      <td>121346350.0</td>\n",
       "      <td>fix-patch</td>\n",
       "      <td>JH636053.3</td>\n",
       "      <td>NW_003871056.3</td>\n",
       "      <td>PATCHES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1Q21</td>\n",
       "      <td>1</td>\n",
       "      <td>143871003.0</td>\n",
       "      <td>150440213.0</td>\n",
       "      <td>fix-patch</td>\n",
       "      <td>JH636052.4</td>\n",
       "      <td>NW_003871055.3</td>\n",
       "      <td>PATCHES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>REGION2</td>\n",
       "      <td>1</td>\n",
       "      <td>153673007.0</td>\n",
       "      <td>153838214.0</td>\n",
       "      <td>novel-patch</td>\n",
       "      <td>GL383518.1</td>\n",
       "      <td>NW_003315905.1</td>\n",
       "      <td>PATCHES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>212</th>\n",
       "      <td>REGION97</td>\n",
       "      <td>X</td>\n",
       "      <td>142299066.0</td>\n",
       "      <td>150174083.0</td>\n",
       "      <td>fix-patch</td>\n",
       "      <td>JH806600.2</td>\n",
       "      <td>NW_004070890.2</td>\n",
       "      <td>PATCHES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213</th>\n",
       "      <td>REGION67</td>\n",
       "      <td>X</td>\n",
       "      <td>151734490.0</td>\n",
       "      <td>154906585.0</td>\n",
       "      <td>fix-patch</td>\n",
       "      <td>JH159150.3</td>\n",
       "      <td>NW_003871103.3</td>\n",
       "      <td>PATCHES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>214</th>\n",
       "      <td>PAR-2</td>\n",
       "      <td>X</td>\n",
       "      <td>154931044.0</td>\n",
       "      <td>155260560.0</td>\n",
       "      <td>PAR</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>215</th>\n",
       "      <td>PAR-1</td>\n",
       "      <td>Y</td>\n",
       "      <td>10001.0</td>\n",
       "      <td>2649520.0</td>\n",
       "      <td>PAR</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>216</th>\n",
       "      <td>PAR-2</td>\n",
       "      <td>Y</td>\n",
       "      <td>59034050.0</td>\n",
       "      <td>59363566.0</td>\n",
       "      <td>PAR</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>217 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         name chr        start          end         type     genbank  \\\n",
       "0    REGION77   1    3823584.0    3845268.0    fix-patch  JH806574.2   \n",
       "1    REGION60   1   31872759.0   32017063.0    fix-patch  GL949741.1   \n",
       "2        1P11   1  120471050.0  121346350.0    fix-patch  JH636053.3   \n",
       "3        1Q21   1  143871003.0  150440213.0    fix-patch  JH636052.4   \n",
       "4     REGION2   1  153673007.0  153838214.0  novel-patch  GL383518.1   \n",
       "..        ...  ..          ...          ...          ...         ...   \n",
       "212  REGION97   X  142299066.0  150174083.0    fix-patch  JH806600.2   \n",
       "213  REGION67   X  151734490.0  154906585.0    fix-patch  JH159150.3   \n",
       "214     PAR-2   X  154931044.0  155260560.0          PAR          na   \n",
       "215     PAR-1   Y      10001.0    2649520.0          PAR          na   \n",
       "216     PAR-2   Y   59034050.0   59363566.0          PAR          na   \n",
       "\n",
       "             refseq     unit  \n",
       "0    NW_004070864.2  PATCHES  \n",
       "1    NW_003571030.1  PATCHES  \n",
       "2    NW_003871056.3  PATCHES  \n",
       "3    NW_003871055.3  PATCHES  \n",
       "4    NW_003315905.1  PATCHES  \n",
       "..              ...      ...  \n",
       "212  NW_004070890.2  PATCHES  \n",
       "213  NW_003871103.3  PATCHES  \n",
       "214              na       na  \n",
       "215              na       na  \n",
       "216              na       na  \n",
       "\n",
       "[217 rows x 8 columns]"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "patches.columns = ['name', 'chr', 'start', 'end', 'type', 'genbank', 'refseq', 'unit']\n",
    "patches"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "993ce6a6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>chr</th>\n",
       "      <th>type.molecule</th>\n",
       "      <th>genbank</th>\n",
       "      <th>relation</th>\n",
       "      <th>refseq</th>\n",
       "      <th>unit</th>\n",
       "      <th>length</th>\n",
       "      <th>ucsc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>1</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000663.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000001.10</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>249250621</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>2</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000664.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000002.11</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>243199373</td>\n",
       "      <td>chr2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>3</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000665.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000003.11</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>198022430</td>\n",
       "      <td>chr3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>4</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000666.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000004.11</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>191154276</td>\n",
       "      <td>chr4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>5</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000667.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000005.9</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>180915260</td>\n",
       "      <td>chr5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>292</th>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>6</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>GL000255.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_167248.1</td>\n",
       "      <td>ALT_REF_LOCI_6</td>\n",
       "      <td>4611984</td>\n",
       "      <td>chr6_qbl_hap6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>293</th>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>6</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>GL000256.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_167249.1</td>\n",
       "      <td>ALT_REF_LOCI_7</td>\n",
       "      <td>4928567</td>\n",
       "      <td>chr6_ssto_hap7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>294</th>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>4</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>GL000257.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_167250.1</td>\n",
       "      <td>ALT_REF_LOCI_8</td>\n",
       "      <td>590426</td>\n",
       "      <td>chr4_ctg9_hap1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>295</th>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>17</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>GL000258.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_167251.1</td>\n",
       "      <td>ALT_REF_LOCI_9</td>\n",
       "      <td>1680828</td>\n",
       "      <td>chr17_ctg5_hap1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>296</th>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>MT</td>\n",
       "      <td>Mitochondrion</td>\n",
       "      <td>J01415.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_012920.1</td>\n",
       "      <td>non-nuclear</td>\n",
       "      <td>16569</td>\n",
       "      <td>chrM</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>297 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   type chr  type.molecule     genbank relation        refseq  \\\n",
       "0    assembled-molecule   1     Chromosome  CM000663.1        =  NC_000001.10   \n",
       "1    assembled-molecule   2     Chromosome  CM000664.1        =  NC_000002.11   \n",
       "2    assembled-molecule   3     Chromosome  CM000665.1        =  NC_000003.11   \n",
       "3    assembled-molecule   4     Chromosome  CM000666.1        =  NC_000004.11   \n",
       "4    assembled-molecule   5     Chromosome  CM000667.1        =   NC_000005.9   \n",
       "..                  ...  ..            ...         ...      ...           ...   \n",
       "292        alt-scaffold   6     Chromosome  GL000255.1        =   NT_167248.1   \n",
       "293        alt-scaffold   6     Chromosome  GL000256.1        =   NT_167249.1   \n",
       "294        alt-scaffold   4     Chromosome  GL000257.1        =   NT_167250.1   \n",
       "295        alt-scaffold  17     Chromosome  GL000258.1        =   NT_167251.1   \n",
       "296  assembled-molecule  MT  Mitochondrion    J01415.2        =   NC_012920.1   \n",
       "\n",
       "                 unit     length             ucsc  \n",
       "0    Primary Assembly  249250621             chr1  \n",
       "1    Primary Assembly  243199373             chr2  \n",
       "2    Primary Assembly  198022430             chr3  \n",
       "3    Primary Assembly  191154276             chr4  \n",
       "4    Primary Assembly  180915260             chr5  \n",
       "..                ...        ...              ...  \n",
       "292    ALT_REF_LOCI_6    4611984    chr6_qbl_hap6  \n",
       "293    ALT_REF_LOCI_7    4928567   chr6_ssto_hap7  \n",
       "294    ALT_REF_LOCI_8     590426   chr4_ctg9_hap1  \n",
       "295    ALT_REF_LOCI_9    1680828  chr17_ctg5_hap1  \n",
       "296       non-nuclear      16569             chrM  \n",
       "\n",
       "[297 rows x 9 columns]"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "regions.columns = [\n",
    "    'name', 'type', 'chr', 'type.molecule', 'genbank', 'relation', 'refseq', 'unit', 'length', 'ucsc'\n",
    "]\n",
    "\n",
    "regions.iloc[:, 1:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "278bc796",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "type\n",
       "fix-patch               131\n",
       "novel-patch              73\n",
       "unplaced-scaffold        39\n",
       "assembled-molecule       25\n",
       "unlocalized-scaffold     20\n",
       "alt-scaffold              9\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "regions['type'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "b8921f43",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "type\n",
       "fix-patch       131\n",
       "novel-patch      73\n",
       "alt-scaffold      9\n",
       "PAR               3\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "patches['type'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d9fbc9e6",
   "metadata": {},
   "source": [
    "We will organize the genomic fragments into chromosomes (with alternative scaffolds, fix patches and \n",
    "novel patches and gaps included). This will produce a comprehensive mapping to the assembly."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58744ae2",
   "metadata": {},
   "source": [
    "### Retrieve primary regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "8ad3958c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>type</th>\n",
       "      <th>chr</th>\n",
       "      <th>genbank</th>\n",
       "      <th>refseq</th>\n",
       "      <th>unit</th>\n",
       "      <th>length.real</th>\n",
       "      <th>ucsc</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>length.map</th>\n",
       "      <th>subtype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>primary</td>\n",
       "      <td>1</td>\n",
       "      <td>CM000663.1</td>\n",
       "      <td>NC_000001.10</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>249250621</td>\n",
       "      <td>chr1</td>\n",
       "      <td>0</td>\n",
       "      <td>249250621</td>\n",
       "      <td>249250621</td>\n",
       "      <td>chromosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>primary</td>\n",
       "      <td>2</td>\n",
       "      <td>CM000664.1</td>\n",
       "      <td>NC_000002.11</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>243199373</td>\n",
       "      <td>chr2</td>\n",
       "      <td>0</td>\n",
       "      <td>243199373</td>\n",
       "      <td>243199373</td>\n",
       "      <td>chromosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>primary</td>\n",
       "      <td>3</td>\n",
       "      <td>CM000665.1</td>\n",
       "      <td>NC_000003.11</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>198022430</td>\n",
       "      <td>chr3</td>\n",
       "      <td>0</td>\n",
       "      <td>198022430</td>\n",
       "      <td>198022430</td>\n",
       "      <td>chromosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>primary</td>\n",
       "      <td>4</td>\n",
       "      <td>CM000666.1</td>\n",
       "      <td>NC_000004.11</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>191154276</td>\n",
       "      <td>chr4</td>\n",
       "      <td>0</td>\n",
       "      <td>191154276</td>\n",
       "      <td>191154276</td>\n",
       "      <td>chromosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>primary</td>\n",
       "      <td>5</td>\n",
       "      <td>CM000667.1</td>\n",
       "      <td>NC_000005.9</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>180915260</td>\n",
       "      <td>chr5</td>\n",
       "      <td>0</td>\n",
       "      <td>180915260</td>\n",
       "      <td>180915260</td>\n",
       "      <td>chromosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>6</td>\n",
       "      <td>primary</td>\n",
       "      <td>6</td>\n",
       "      <td>CM000668.1</td>\n",
       "      <td>NC_000006.11</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>171115067</td>\n",
       "      <td>chr6</td>\n",
       "      <td>0</td>\n",
       "      <td>171115067</td>\n",
       "      <td>171115067</td>\n",
       "      <td>chromosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7</td>\n",
       "      <td>primary</td>\n",
       "      <td>7</td>\n",
       "      <td>CM000669.1</td>\n",
       "      <td>NC_000007.13</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>159138663</td>\n",
       "      <td>chr7</td>\n",
       "      <td>0</td>\n",
       "      <td>159138663</td>\n",
       "      <td>159138663</td>\n",
       "      <td>chromosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>8</td>\n",
       "      <td>primary</td>\n",
       "      <td>8</td>\n",
       "      <td>CM000670.1</td>\n",
       "      <td>NC_000008.10</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>146364022</td>\n",
       "      <td>chr8</td>\n",
       "      <td>0</td>\n",
       "      <td>146364022</td>\n",
       "      <td>146364022</td>\n",
       "      <td>chromosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>9</td>\n",
       "      <td>primary</td>\n",
       "      <td>9</td>\n",
       "      <td>CM000671.1</td>\n",
       "      <td>NC_000009.11</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>141213431</td>\n",
       "      <td>chr9</td>\n",
       "      <td>0</td>\n",
       "      <td>141213431</td>\n",
       "      <td>141213431</td>\n",
       "      <td>chromosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>10</td>\n",
       "      <td>primary</td>\n",
       "      <td>10</td>\n",
       "      <td>CM000672.1</td>\n",
       "      <td>NC_000010.10</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>135534747</td>\n",
       "      <td>chr10</td>\n",
       "      <td>0</td>\n",
       "      <td>135534747</td>\n",
       "      <td>135534747</td>\n",
       "      <td>chromosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>11</td>\n",
       "      <td>primary</td>\n",
       "      <td>11</td>\n",
       "      <td>CM000673.1</td>\n",
       "      <td>NC_000011.9</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>135006516</td>\n",
       "      <td>chr11</td>\n",
       "      <td>0</td>\n",
       "      <td>135006516</td>\n",
       "      <td>135006516</td>\n",
       "      <td>chromosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>12</td>\n",
       "      <td>primary</td>\n",
       "      <td>12</td>\n",
       "      <td>CM000674.1</td>\n",
       "      <td>NC_000012.11</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>133851895</td>\n",
       "      <td>chr12</td>\n",
       "      <td>0</td>\n",
       "      <td>133851895</td>\n",
       "      <td>133851895</td>\n",
       "      <td>chromosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>13</td>\n",
       "      <td>primary</td>\n",
       "      <td>13</td>\n",
       "      <td>CM000675.1</td>\n",
       "      <td>NC_000013.10</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>115169878</td>\n",
       "      <td>chr13</td>\n",
       "      <td>0</td>\n",
       "      <td>115169878</td>\n",
       "      <td>115169878</td>\n",
       "      <td>chromosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>14</td>\n",
       "      <td>primary</td>\n",
       "      <td>14</td>\n",
       "      <td>CM000676.1</td>\n",
       "      <td>NC_000014.8</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>107349540</td>\n",
       "      <td>chr14</td>\n",
       "      <td>0</td>\n",
       "      <td>107349540</td>\n",
       "      <td>107349540</td>\n",
       "      <td>chromosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>15</td>\n",
       "      <td>primary</td>\n",
       "      <td>15</td>\n",
       "      <td>CM000677.1</td>\n",
       "      <td>NC_000015.9</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>102531392</td>\n",
       "      <td>chr15</td>\n",
       "      <td>0</td>\n",
       "      <td>102531392</td>\n",
       "      <td>102531392</td>\n",
       "      <td>chromosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>16</td>\n",
       "      <td>primary</td>\n",
       "      <td>16</td>\n",
       "      <td>CM000678.1</td>\n",
       "      <td>NC_000016.9</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>90354753</td>\n",
       "      <td>chr16</td>\n",
       "      <td>0</td>\n",
       "      <td>90354753</td>\n",
       "      <td>90354753</td>\n",
       "      <td>chromosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>17</td>\n",
       "      <td>primary</td>\n",
       "      <td>17</td>\n",
       "      <td>CM000679.1</td>\n",
       "      <td>NC_000017.10</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>81195210</td>\n",
       "      <td>chr17</td>\n",
       "      <td>0</td>\n",
       "      <td>81195210</td>\n",
       "      <td>81195210</td>\n",
       "      <td>chromosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>18</td>\n",
       "      <td>primary</td>\n",
       "      <td>18</td>\n",
       "      <td>CM000680.1</td>\n",
       "      <td>NC_000018.9</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>78077248</td>\n",
       "      <td>chr18</td>\n",
       "      <td>0</td>\n",
       "      <td>78077248</td>\n",
       "      <td>78077248</td>\n",
       "      <td>chromosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>19</td>\n",
       "      <td>primary</td>\n",
       "      <td>19</td>\n",
       "      <td>CM000681.1</td>\n",
       "      <td>NC_000019.9</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>59128983</td>\n",
       "      <td>chr19</td>\n",
       "      <td>0</td>\n",
       "      <td>59128983</td>\n",
       "      <td>59128983</td>\n",
       "      <td>chromosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>20</td>\n",
       "      <td>primary</td>\n",
       "      <td>20</td>\n",
       "      <td>CM000682.1</td>\n",
       "      <td>NC_000020.10</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>63025520</td>\n",
       "      <td>chr20</td>\n",
       "      <td>0</td>\n",
       "      <td>63025520</td>\n",
       "      <td>63025520</td>\n",
       "      <td>chromosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>21</td>\n",
       "      <td>primary</td>\n",
       "      <td>21</td>\n",
       "      <td>CM000683.1</td>\n",
       "      <td>NC_000021.8</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>48129895</td>\n",
       "      <td>chr21</td>\n",
       "      <td>0</td>\n",
       "      <td>48129895</td>\n",
       "      <td>48129895</td>\n",
       "      <td>chromosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>22</td>\n",
       "      <td>primary</td>\n",
       "      <td>22</td>\n",
       "      <td>CM000684.1</td>\n",
       "      <td>NC_000022.10</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>51304566</td>\n",
       "      <td>chr22</td>\n",
       "      <td>0</td>\n",
       "      <td>51304566</td>\n",
       "      <td>51304566</td>\n",
       "      <td>chromosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>X</td>\n",
       "      <td>primary</td>\n",
       "      <td>X</td>\n",
       "      <td>CM000685.1</td>\n",
       "      <td>NC_000023.10</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>155270560</td>\n",
       "      <td>chrX</td>\n",
       "      <td>0</td>\n",
       "      <td>155270560</td>\n",
       "      <td>155270560</td>\n",
       "      <td>chromosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Y</td>\n",
       "      <td>primary</td>\n",
       "      <td>Y</td>\n",
       "      <td>CM000686.1</td>\n",
       "      <td>NC_000024.9</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>59373566</td>\n",
       "      <td>chrY</td>\n",
       "      <td>0</td>\n",
       "      <td>59373566</td>\n",
       "      <td>59373566</td>\n",
       "      <td>chromosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>296</th>\n",
       "      <td>MT</td>\n",
       "      <td>primary</td>\n",
       "      <td>MT</td>\n",
       "      <td>J01415.2</td>\n",
       "      <td>NC_012920.1</td>\n",
       "      <td>non-nuclear</td>\n",
       "      <td>16569</td>\n",
       "      <td>chrM</td>\n",
       "      <td>0</td>\n",
       "      <td>16569</td>\n",
       "      <td>16569</td>\n",
       "      <td>mitochondrion</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    name     type chr     genbank        refseq              unit  \\\n",
       "0      1  primary   1  CM000663.1  NC_000001.10  Primary Assembly   \n",
       "1      2  primary   2  CM000664.1  NC_000002.11  Primary Assembly   \n",
       "2      3  primary   3  CM000665.1  NC_000003.11  Primary Assembly   \n",
       "3      4  primary   4  CM000666.1  NC_000004.11  Primary Assembly   \n",
       "4      5  primary   5  CM000667.1   NC_000005.9  Primary Assembly   \n",
       "5      6  primary   6  CM000668.1  NC_000006.11  Primary Assembly   \n",
       "6      7  primary   7  CM000669.1  NC_000007.13  Primary Assembly   \n",
       "7      8  primary   8  CM000670.1  NC_000008.10  Primary Assembly   \n",
       "8      9  primary   9  CM000671.1  NC_000009.11  Primary Assembly   \n",
       "9     10  primary  10  CM000672.1  NC_000010.10  Primary Assembly   \n",
       "10    11  primary  11  CM000673.1   NC_000011.9  Primary Assembly   \n",
       "11    12  primary  12  CM000674.1  NC_000012.11  Primary Assembly   \n",
       "12    13  primary  13  CM000675.1  NC_000013.10  Primary Assembly   \n",
       "13    14  primary  14  CM000676.1   NC_000014.8  Primary Assembly   \n",
       "14    15  primary  15  CM000677.1   NC_000015.9  Primary Assembly   \n",
       "15    16  primary  16  CM000678.1   NC_000016.9  Primary Assembly   \n",
       "16    17  primary  17  CM000679.1  NC_000017.10  Primary Assembly   \n",
       "17    18  primary  18  CM000680.1   NC_000018.9  Primary Assembly   \n",
       "18    19  primary  19  CM000681.1   NC_000019.9  Primary Assembly   \n",
       "19    20  primary  20  CM000682.1  NC_000020.10  Primary Assembly   \n",
       "20    21  primary  21  CM000683.1   NC_000021.8  Primary Assembly   \n",
       "21    22  primary  22  CM000684.1  NC_000022.10  Primary Assembly   \n",
       "22     X  primary   X  CM000685.1  NC_000023.10  Primary Assembly   \n",
       "23     Y  primary   Y  CM000686.1   NC_000024.9  Primary Assembly   \n",
       "296   MT  primary  MT    J01415.2   NC_012920.1       non-nuclear   \n",
       "\n",
       "     length.real   ucsc  start        end  length.map        subtype  \n",
       "0      249250621   chr1      0  249250621   249250621     chromosome  \n",
       "1      243199373   chr2      0  243199373   243199373     chromosome  \n",
       "2      198022430   chr3      0  198022430   198022430     chromosome  \n",
       "3      191154276   chr4      0  191154276   191154276     chromosome  \n",
       "4      180915260   chr5      0  180915260   180915260     chromosome  \n",
       "5      171115067   chr6      0  171115067   171115067     chromosome  \n",
       "6      159138663   chr7      0  159138663   159138663     chromosome  \n",
       "7      146364022   chr8      0  146364022   146364022     chromosome  \n",
       "8      141213431   chr9      0  141213431   141213431     chromosome  \n",
       "9      135534747  chr10      0  135534747   135534747     chromosome  \n",
       "10     135006516  chr11      0  135006516   135006516     chromosome  \n",
       "11     133851895  chr12      0  133851895   133851895     chromosome  \n",
       "12     115169878  chr13      0  115169878   115169878     chromosome  \n",
       "13     107349540  chr14      0  107349540   107349540     chromosome  \n",
       "14     102531392  chr15      0  102531392   102531392     chromosome  \n",
       "15      90354753  chr16      0   90354753    90354753     chromosome  \n",
       "16      81195210  chr17      0   81195210    81195210     chromosome  \n",
       "17      78077248  chr18      0   78077248    78077248     chromosome  \n",
       "18      59128983  chr19      0   59128983    59128983     chromosome  \n",
       "19      63025520  chr20      0   63025520    63025520     chromosome  \n",
       "20      48129895  chr21      0   48129895    48129895     chromosome  \n",
       "21      51304566  chr22      0   51304566    51304566     chromosome  \n",
       "22     155270560   chrX      0  155270560   155270560     chromosome  \n",
       "23      59373566   chrY      0   59373566    59373566     chromosome  \n",
       "296        16569   chrM      0      16569       16569  mitochondrion  "
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "primary = regions.loc[regions['type'] == 'assembled-molecule', :].copy()\n",
    "primary['start'] = 0\n",
    "primary['end'] = primary['length']\n",
    "primary['length.map'] = primary['length']\n",
    "primary['type'] = 'primary'\n",
    "primary['subtype'] = primary['type.molecule'].str.lower()\n",
    "del primary['type.molecule']\n",
    "primary = primary.rename(columns = {'length': 'length.real'})\n",
    "del primary['relation']\n",
    "primary"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6c8c017a",
   "metadata": {},
   "source": [
    "### Patches and alternative scaffolds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "1fa3f908",
   "metadata": {},
   "outputs": [],
   "source": [
    "aligned = patches[['name', 'chr', 'start', 'end', 'genbank', 'refseq', 'unit', 'type']] \\\n",
    "    .join(regions[['genbank', 'length']].set_index('genbank'), on = 'genbank', how = 'left')\n",
    "aligned = aligned.loc[[x in primary.chr.tolist() for x in aligned['chr']]]\n",
    "aligned['length.real'] = aligned['length']\n",
    "aligned['length.map'] = aligned['end'] - aligned['start']\n",
    "aligned['type'] = [x.lower() for x in aligned['type']]\n",
    "# aligned['subtype'] = aligned['type']\n",
    "del aligned['length']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "8f6e717b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>genbank</th>\n",
       "      <th>refseq</th>\n",
       "      <th>unit</th>\n",
       "      <th>type</th>\n",
       "      <th>length.real</th>\n",
       "      <th>length.map</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>REGION77</td>\n",
       "      <td>1</td>\n",
       "      <td>3823584.0</td>\n",
       "      <td>3845268.0</td>\n",
       "      <td>JH806574.2</td>\n",
       "      <td>NW_004070864.2</td>\n",
       "      <td>PATCHES</td>\n",
       "      <td>fix-patch</td>\n",
       "      <td>22982.0</td>\n",
       "      <td>21684.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>REGION60</td>\n",
       "      <td>1</td>\n",
       "      <td>31872759.0</td>\n",
       "      <td>32017063.0</td>\n",
       "      <td>GL949741.1</td>\n",
       "      <td>NW_003571030.1</td>\n",
       "      <td>PATCHES</td>\n",
       "      <td>fix-patch</td>\n",
       "      <td>151551.0</td>\n",
       "      <td>144304.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1P11</td>\n",
       "      <td>1</td>\n",
       "      <td>120471050.0</td>\n",
       "      <td>121346350.0</td>\n",
       "      <td>JH636053.3</td>\n",
       "      <td>NW_003871056.3</td>\n",
       "      <td>PATCHES</td>\n",
       "      <td>fix-patch</td>\n",
       "      <td>1676126.0</td>\n",
       "      <td>875300.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1Q21</td>\n",
       "      <td>1</td>\n",
       "      <td>143871003.0</td>\n",
       "      <td>150440213.0</td>\n",
       "      <td>JH636052.4</td>\n",
       "      <td>NW_003871055.3</td>\n",
       "      <td>PATCHES</td>\n",
       "      <td>fix-patch</td>\n",
       "      <td>7283150.0</td>\n",
       "      <td>6569210.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>REGION2</td>\n",
       "      <td>1</td>\n",
       "      <td>153673007.0</td>\n",
       "      <td>153838214.0</td>\n",
       "      <td>GL383518.1</td>\n",
       "      <td>NW_003315905.1</td>\n",
       "      <td>PATCHES</td>\n",
       "      <td>novel-patch</td>\n",
       "      <td>182439.0</td>\n",
       "      <td>165207.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>212</th>\n",
       "      <td>REGION97</td>\n",
       "      <td>X</td>\n",
       "      <td>142299066.0</td>\n",
       "      <td>150174083.0</td>\n",
       "      <td>JH806600.2</td>\n",
       "      <td>NW_004070890.2</td>\n",
       "      <td>PATCHES</td>\n",
       "      <td>fix-patch</td>\n",
       "      <td>6530008.0</td>\n",
       "      <td>7875017.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213</th>\n",
       "      <td>REGION67</td>\n",
       "      <td>X</td>\n",
       "      <td>151734490.0</td>\n",
       "      <td>154906585.0</td>\n",
       "      <td>JH159150.3</td>\n",
       "      <td>NW_003871103.3</td>\n",
       "      <td>PATCHES</td>\n",
       "      <td>fix-patch</td>\n",
       "      <td>3110903.0</td>\n",
       "      <td>3172095.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>214</th>\n",
       "      <td>PAR-2</td>\n",
       "      <td>X</td>\n",
       "      <td>154931044.0</td>\n",
       "      <td>155260560.0</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>par</td>\n",
       "      <td>NaN</td>\n",
       "      <td>329516.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>215</th>\n",
       "      <td>PAR-1</td>\n",
       "      <td>Y</td>\n",
       "      <td>10001.0</td>\n",
       "      <td>2649520.0</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>par</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2639519.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>216</th>\n",
       "      <td>PAR-2</td>\n",
       "      <td>Y</td>\n",
       "      <td>59034050.0</td>\n",
       "      <td>59363566.0</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>par</td>\n",
       "      <td>NaN</td>\n",
       "      <td>329516.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>216 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         name chr        start          end     genbank          refseq  \\\n",
       "0    REGION77   1    3823584.0    3845268.0  JH806574.2  NW_004070864.2   \n",
       "1    REGION60   1   31872759.0   32017063.0  GL949741.1  NW_003571030.1   \n",
       "2        1P11   1  120471050.0  121346350.0  JH636053.3  NW_003871056.3   \n",
       "3        1Q21   1  143871003.0  150440213.0  JH636052.4  NW_003871055.3   \n",
       "4     REGION2   1  153673007.0  153838214.0  GL383518.1  NW_003315905.1   \n",
       "..        ...  ..          ...          ...         ...             ...   \n",
       "212  REGION97   X  142299066.0  150174083.0  JH806600.2  NW_004070890.2   \n",
       "213  REGION67   X  151734490.0  154906585.0  JH159150.3  NW_003871103.3   \n",
       "214     PAR-2   X  154931044.0  155260560.0          na              na   \n",
       "215     PAR-1   Y      10001.0    2649520.0          na              na   \n",
       "216     PAR-2   Y   59034050.0   59363566.0          na              na   \n",
       "\n",
       "        unit         type  length.real  length.map  \n",
       "0    PATCHES    fix-patch      22982.0     21684.0  \n",
       "1    PATCHES    fix-patch     151551.0    144304.0  \n",
       "2    PATCHES    fix-patch    1676126.0    875300.0  \n",
       "3    PATCHES    fix-patch    7283150.0   6569210.0  \n",
       "4    PATCHES  novel-patch     182439.0    165207.0  \n",
       "..       ...          ...          ...         ...  \n",
       "212  PATCHES    fix-patch    6530008.0   7875017.0  \n",
       "213  PATCHES    fix-patch    3110903.0   3172095.0  \n",
       "214       na          par          NaN    329516.0  \n",
       "215       na          par          NaN   2639519.0  \n",
       "216       na          par          NaN    329516.0  \n",
       "\n",
       "[216 rows x 10 columns]"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aligned"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c140fb39",
   "metadata": {},
   "source": [
    "### Gaps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "9cda5ee5",
   "metadata": {},
   "outputs": [],
   "source": [
    "gps = gaps.join(primary[['refseq', 'chr', 'ucsc']].set_index('refseq'), on = '.seqid', how = 'left')\n",
    "gps = gps[~ gps.chr.isna()]\n",
    "gps['name'] = ['gap:' + str(i + 1) for i in range(len(gps))]\n",
    "gps['subtype'] = [x.replace('_', '-') for x in gps['type']]\n",
    "gps['type'] = 'gap'\n",
    "del gps['linkage.evidence']\n",
    "gps = gps.rename(columns = {'.seqid': 'refseq'})\n",
    "gps['length.real'] = gps['length']\n",
    "gps['length.map'] = gps['length']\n",
    "del gps['length']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "b2b41bc6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>refseq</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>type</th>\n",
       "      <th>chr</th>\n",
       "      <th>ucsc</th>\n",
       "      <th>name</th>\n",
       "      <th>subtype</th>\n",
       "      <th>length.real</th>\n",
       "      <th>length.map</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>NC_000001.10</td>\n",
       "      <td>1</td>\n",
       "      <td>10000</td>\n",
       "      <td>gap</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>gap:1</td>\n",
       "      <td>telomere</td>\n",
       "      <td>10000</td>\n",
       "      <td>10000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>NC_000001.10</td>\n",
       "      <td>177418</td>\n",
       "      <td>227417</td>\n",
       "      <td>gap</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>gap:2</td>\n",
       "      <td>within-scaffold</td>\n",
       "      <td>50000</td>\n",
       "      <td>50000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>NC_000001.10</td>\n",
       "      <td>267720</td>\n",
       "      <td>317719</td>\n",
       "      <td>gap</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>gap:3</td>\n",
       "      <td>between-scaffolds</td>\n",
       "      <td>50000</td>\n",
       "      <td>50000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>NC_000001.10</td>\n",
       "      <td>471369</td>\n",
       "      <td>521368</td>\n",
       "      <td>gap</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>gap:4</td>\n",
       "      <td>between-scaffolds</td>\n",
       "      <td>50000</td>\n",
       "      <td>50000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>NC_000001.10</td>\n",
       "      <td>2634221</td>\n",
       "      <td>2684220</td>\n",
       "      <td>gap</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>gap:5</td>\n",
       "      <td>within-scaffold</td>\n",
       "      <td>50000</td>\n",
       "      <td>50000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>508</th>\n",
       "      <td>NC_000024.9</td>\n",
       "      <td>22369680</td>\n",
       "      <td>22419679</td>\n",
       "      <td>gap</td>\n",
       "      <td>Y</td>\n",
       "      <td>chrY</td>\n",
       "      <td>gap:366</td>\n",
       "      <td>within-scaffold</td>\n",
       "      <td>50000</td>\n",
       "      <td>50000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>509</th>\n",
       "      <td>NC_000024.9</td>\n",
       "      <td>23901429</td>\n",
       "      <td>23951428</td>\n",
       "      <td>gap</td>\n",
       "      <td>Y</td>\n",
       "      <td>chrY</td>\n",
       "      <td>gap:367</td>\n",
       "      <td>between-scaffolds</td>\n",
       "      <td>50000</td>\n",
       "      <td>50000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>510</th>\n",
       "      <td>NC_000024.9</td>\n",
       "      <td>28819362</td>\n",
       "      <td>58819361</td>\n",
       "      <td>gap</td>\n",
       "      <td>Y</td>\n",
       "      <td>chrY</td>\n",
       "      <td>gap:368</td>\n",
       "      <td>heterochromatin</td>\n",
       "      <td>30000000</td>\n",
       "      <td>30000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>511</th>\n",
       "      <td>NC_000024.9</td>\n",
       "      <td>58917657</td>\n",
       "      <td>58967656</td>\n",
       "      <td>gap</td>\n",
       "      <td>Y</td>\n",
       "      <td>chrY</td>\n",
       "      <td>gap:369</td>\n",
       "      <td>between-scaffolds</td>\n",
       "      <td>50000</td>\n",
       "      <td>50000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>512</th>\n",
       "      <td>NC_000024.9</td>\n",
       "      <td>59363567</td>\n",
       "      <td>59373566</td>\n",
       "      <td>gap</td>\n",
       "      <td>Y</td>\n",
       "      <td>chrY</td>\n",
       "      <td>gap:370</td>\n",
       "      <td>telomere</td>\n",
       "      <td>10000</td>\n",
       "      <td>10000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>370 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           refseq     start       end type chr  ucsc     name  \\\n",
       "143  NC_000001.10         1     10000  gap   1  chr1    gap:1   \n",
       "144  NC_000001.10    177418    227417  gap   1  chr1    gap:2   \n",
       "145  NC_000001.10    267720    317719  gap   1  chr1    gap:3   \n",
       "146  NC_000001.10    471369    521368  gap   1  chr1    gap:4   \n",
       "147  NC_000001.10   2634221   2684220  gap   1  chr1    gap:5   \n",
       "..            ...       ...       ...  ...  ..   ...      ...   \n",
       "508   NC_000024.9  22369680  22419679  gap   Y  chrY  gap:366   \n",
       "509   NC_000024.9  23901429  23951428  gap   Y  chrY  gap:367   \n",
       "510   NC_000024.9  28819362  58819361  gap   Y  chrY  gap:368   \n",
       "511   NC_000024.9  58917657  58967656  gap   Y  chrY  gap:369   \n",
       "512   NC_000024.9  59363567  59373566  gap   Y  chrY  gap:370   \n",
       "\n",
       "               subtype  length.real  length.map  \n",
       "143           telomere        10000       10000  \n",
       "144    within-scaffold        50000       50000  \n",
       "145  between-scaffolds        50000       50000  \n",
       "146  between-scaffolds        50000       50000  \n",
       "147    within-scaffold        50000       50000  \n",
       "..                 ...          ...         ...  \n",
       "508    within-scaffold        50000       50000  \n",
       "509  between-scaffolds        50000       50000  \n",
       "510    heterochromatin     30000000    30000000  \n",
       "511  between-scaffolds        50000       50000  \n",
       "512           telomere        10000       10000  \n",
       "\n",
       "[370 rows x 10 columns]"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gps"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29ba0ea8",
   "metadata": {},
   "source": [
    "### Merge"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "5769a2c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "conc = pd.concat([primary, aligned, gps])\n",
    "conc['ucsc'] = ['chrM' if x == 'MT' else 'chr' + x for x in conc['chr']]\n",
    "conc['genbank'] = [x if str(x) != 'nan' else 'na' for x in conc['genbank']]\n",
    "conc['unit'] = [x if str(x) != 'nan' else 'na' for x in conc['unit']]\n",
    "conc['length.real'] = [x if str(x) != 'nan' else y for x, y in zip(conc['length.real'], conc['length.map'])]\n",
    "conc['subtype'] = [x if str(x) != 'nan' else y for x, y in zip(conc['subtype'], conc['type'])]\n",
    "conc['length.map'] = conc['length.map'].astype('int')\n",
    "conc['start'] = conc['start'].astype('int')\n",
    "conc['end'] = conc['end'].astype('int')\n",
    "conc = conc.sort_values(['chr', 'start', 'length.map'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "12a1f806",
   "metadata": {},
   "outputs": [],
   "source": [
    "conc.to_csv('architecture.tsv', sep = '\\t', index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "f64a0679",
   "metadata": {},
   "outputs": [],
   "source": [
    "sizes = conc.loc[conc.type == 'primary', ['ucsc', 'length.map']].set_index('ucsc').to_dict()['length.map']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "80adac8a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'chr1': 249250621,\n",
       " 'chr10': 135534747,\n",
       " 'chr11': 135006516,\n",
       " 'chr12': 133851895,\n",
       " 'chr13': 115169878,\n",
       " 'chr14': 107349540,\n",
       " 'chr15': 102531392,\n",
       " 'chr16': 90354753,\n",
       " 'chr17': 81195210,\n",
       " 'chr18': 78077248,\n",
       " 'chr19': 59128983,\n",
       " 'chr2': 243199373,\n",
       " 'chr20': 63025520,\n",
       " 'chr21': 48129895,\n",
       " 'chr22': 51304566,\n",
       " 'chr3': 198022430,\n",
       " 'chr4': 191154276,\n",
       " 'chr5': 180915260,\n",
       " 'chr6': 171115067,\n",
       " 'chr7': 159138663,\n",
       " 'chr8': 146364022,\n",
       " 'chr9': 141213431,\n",
       " 'chrM': 16569,\n",
       " 'chrX': 155270560,\n",
       " 'chrY': 59373566}"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sizes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "5542296f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "with open('size.json', 'w') as f:\n",
    "    json.dump(sizes, f, indent = 4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "189cca02",
   "metadata": {},
   "outputs": [],
   "source": [
    "!rm gaps.tsv patches.tsv regions.tsv"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "exprmat.py310.r43.torch",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
